Amend cloudprovider metric proposal for new metric names

gnufied opened the pull request:
We are going to use same name for all metrics. But each api request
being made will be added as a dimension.

This has the benefit that - aggregating metrics is way easier without
without resorting to wildcards and stuff.

xref - https://github.com/kubernetes/features/issues/182

cc @kubernetes/sig-storage-pr-reviews @saad-ali @piosz @brancz @fabxc

gnufied commented:
cc @justinsb

brancz commented:
I think it is now less clear what the namespace stands for, can you elaborate in the doc?

fabxc commented on contributors/design-proposals/cloudprovider-storage-metrics.md:
```diff
@@ -58,28 +58,39 @@ For GCE we simply use `gensupport.RegisterHook()` to register a function which w
-4. gce_attach_disk
-5. gce_detach_disk
-6. gce_list_disk
+All gce metrics will be named - `gce_api_latency_duration_seconds`. Namespace and api request
```
GCE and API for readability. Why the "–"?

fabxc replied:
Isn't `latency_duration` kind of doubling down on the meaning? Seems either of them would be enough but `request` should be in there somehow. 

How about `gce_api_request_duration_seconds`?

fabxc commented on contributors/design-proposals/cloudprovider-storage-metrics.md:
````diff
@@ -58,28 +58,39 @@ For GCE we simply use `gensupport.RegisterHook()` to register a function which w
+The complete list of time series will look like:
+
+```
+gce_api_latency_duration_seconds { namespace: "", api_request: "instance_list"}
````
The `:` should be `=` if we follow the format of the /metrics endpoints.

gnufied commented:
@brancz @fabxc namespace is kubernetes namespace. Now that I think of, we can explicitly name dimension to be `kube_namespace` so as to avoid any confusion

brancz commented:
I understand it's the kubernetes namespace, but I'm wondering why this information is necessary at all. If we provide a namespace, when we should also provide the object this request was made for, no?

gnufied commented:
In a multi tenant environment there is a desire to calculate which namespace is using most api requests.  This way we can identify namespaces which are using the cloudprovider api most. 

To be fair namespace information is not always available but my idea is to emit that information whenever it is available.

I can amend the metrics to emit object for which api request is being made for. But similar to namespaces - we do not always have that information available at call site where metrics are created.  Moreover - sometimes a api call can be made against 2 different types of objects and we can't uniquely identify against which object the api request was made (`describeVolume` for example). To capture this information - we have to make larger refactor of cloudprovider code which I am hesitant to do at this point.

fabxc commented:
That's fine. But if the info is not available it should be filled with a sentinel value rather than the empty string, e.g. `namespace="<unknown>"`.

brancz commented:
Makes sense. To stay conform to the instrumentation guide for kubernetes components we should definitely stick to `namespace`, and @fabxc's comment also makes sense, we've done something similar in other components, such as the rescheduler.

gnufied commented:
I have updated the design proposal as per latest discussion me, @justinsb and @bowei had. I have dropped namespace as a dimension. I have also added bit about cloudprovider maintainer adding dimension as needed.

bowei commented on contributors/design-proposals/cloudprovider-storage-metrics.md:
```diff
@@ -49,37 +49,46 @@ Since we are interested in count(or rate) and latency percentile metrics of API
+`Histogram` - both count and latency will be automatically calculated.
+
+All gce metrics will be named - `cloudprovider_gce_api_duration_seconds`. api request
+being made will be reported as dimensions. The cloudprovider maintainer may choose to add additional dimensions
```
This statement seems like it should be not specific to GCE?

bowei commented on contributors/design-proposals/cloudprovider-storage-metrics.md:
```diff
@@ -49,37 +49,46 @@ Since we are interested in count(or rate) and latency percentile metrics of API
+being made will be reported as dimensions. The cloudprovider maintainer may choose to add additional dimensions
+as needed.
+
+The complete list of time series will look like:
```
Remove "complete"? This is not a complete list, it's a list of examples? It'll make the document hard to maintain if we have to keep it in sync with the code.

bowei commented on contributors/design-proposals/cloudprovider-storage-metrics.md:
```diff
@@ -49,37 +49,46 @@ Since we are interested in count(or rate) and latency percentile metrics of API
 
-We will be using `HistogramVec` type so as we can attach dimensions at runtime. Whenever available
-`namespace` will reported as a dimension with the metric.
+We will be using `HistogramVec` type so as we can attach dimensions at runtime.
```
Should we almost keep track of errors? E.g. the error count in the GCE proposed implementation...

gnufied replied:
updated.

gnufied commented:
@bowei I have updated proposal to include error counter.

bowei commented on contributors/design-proposals/cloudprovider-storage-metrics.md:
```diff
@@ -49,37 +49,87 @@ Since we are interested in count(or rate) and latency percentile metrics of API
+
+#### GCE Latency metrics
+
+All gce latency metrics will be named - `cloudprovider_gce_api_duration_seconds`. api request
```
You will need to add the naming for the error counter as well

cloudprovider_gce_api_errors

bowei replied:
whoops ignore this comment

bowei replied:
Now thinking about it, we may not want all of the metrics to be put into the same name, as this makes all of the metrics need to share the same dimensions (e.g. zone, region, etc). The dimensions may not make sense for most of the metrics.

For example, if you look at my PR, disk has (zone, node), but other metrics only have (zone) or ().

It probably would better to have:
cloudprovider_{gce,aws, ...}_{disk,loadbalancer, ...}

to avoid having too many unused properties.

What do you think?

gnufied replied:
What other properties we are considering btw? I was thinking these metrics are somewhat lower level and meant to only track API latency and rate/quota usage.

If we are going to need to classify the metrics between `disk` or `loadbalancer`  or something else, it might be better to use API request name with metrics itself rather than moving it to dimension. It is needless mental overload having to classify the metrics.

gnufied replied:
We spoke about this offline and we agreed to use same metric name for all metrics.  A cloudprovider may choose to emit a dimension's value as `n/a` if it is not available at metric emission point.

bowei commented on contributors/design-proposals/cloudprovider-storage-metrics.md:
````diff
@@ -49,37 +49,87 @@ Since we are interested in count(or rate) and latency percentile metrics of API
+To begin we will start emitting following metrics:
+
+```
+cloudprovider_gce_api_duration_seconds { api_request = "instance_list"}
````
api_ in `api_request` is redundant (the metric name already has `api` in it)

gnufied replied:
lets just call it just `request` ? I don't have strong opinions on this one.

bowei replied:
request works for me

gnufied replied:
renamed

gnufied commented:
@piosz @fabxc  PTAL. Since last update - have made following changes to the proposal:

1. Use same metric name for all cloudprovider metrics from same cloudprovider. Use `request` as a dimension to determine API action being taken.
2. Drop `namespace` as a requirement from cloudprovider metrics. The only dimension which will be always present is `request` dimension. Beyond that, a cloudprovider may choose to emit any dimensions they need.
3. If a dimension's value is not available (and declared metric defines the dimension). Its value can be set to `n/a` during the emission.

fabxc commented on contributors/design-proposals/cloudprovider-storage-metrics.md:
````diff
@@ -49,37 +49,88 @@ Since we are interested in count(or rate) and latency percentile metrics of API
+To begin we will start emitting following metrics:
+
+```
+cloudprovider_gce_api_duration_seconds { request = "instance_list"}
````
I'd tend to include make the suffix `request_duration_seconds` regardless. It's not a super hard requirement but certainly the common approach. Plus, "API duration" reads kind of awkward – when just looking at the metric name, it's unclear that it's about a request.

gnufied replied:
Do you mean the name ought to be `cloudprovider_gce_request_duration_seconds`  or `cloudprovider_gce_api_request_duration_seconds` ? I was mostly concerned with making the name too big - but I do not have any objections to changing it.

brancz replied:
I agree that it makes the name pretty long, but it is much more understandable what the metric represents so I'd vote for the convention/best practice of `request_duration_seconds`.

gnufied replied:
Not sure I am completely agree with keeping such long names. But in the interest of getting this shipped - I will lay down my guns.

bowei commented:
/LGTM for me

brancz commented:
/lgtm

gnufied commented:
@saad-ali can you approve and merge?
